### PR TITLE
Merge wolfgangs spark code branch to develop

### PR DESCRIPTION
**Pull request by christiam**

### CODE REVIEW

File: pipeline/src/main/java/BC_CONTEXT.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            BC_REQUEST request = BC_REQUEST_READER.parse_from_string( req_string );
+            if ( request != null )
+                res = add_request( request, ps );
+            else if ( ps != null )
+                ps.printf( "invalid request '%s'\n", req_string );
+        }
+        return res;
+    }
+
+	/* return 1..done, 0..not-done, -1..invalid*/
```
Comment (vartanianmh) [inline pipeline/src/main/java/BC_CONTEXT.java:105]: enum?

File: pipeline/src/main/java/BC_GCP_TOOLS.java [hunk 2]
```diff
@@ -150,53 +152,85 @@ private Integer list_bucket( final String bucket, List< String > lst )
...
-        try
-        {
-            Storage.Objects.Get obj = storage.objects().get( bucket, key );
-            if ( obj != null )
-            {
-                obj.getMediaHttpDownloader().setDirectDownloadEnabled( true );
+        boolean res = BC_UTILS.create_paths_if_neccessary( dst_filename );
+		if ( res )
+		{
+			File f_lock = new File( String.format( "%s.lock", dst_filename ) );
```
Comment (vartanianmh) [inline pipeline/src/main/java/BC_GCP_TOOLS.java:178]: Looks a little racy. Is there a Java lockfile API?

**Review by vartanianmh — COMMENTED** [review]
Good for now

**Review by vartanianmh — APPROVED** [review]

**Review by vartanianmh — APPROVED** [review]

File: pipeline/src/main/java/BC_GCP_TOOLS.java (hunk 2 above)
Comment (christiam) [reply pipeline/src/main/java/BC_GCP_TOOLS.java:178]: [FileLock ](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileLock.html) comes to mind. FWIW, I was mostly focused in getting the latest code changes merged into the main trunk.

Is the code ready for a code review? @wraetz

File: pipeline/src/main/java/BC_CONTEXT.java (hunk 1 above)
Comment (christiam) [reply pipeline/src/main/java/BC_CONTEXT.java:105]: Sounds reasonable, though I'd prefer an exception on failure.

**Pull request merged**